ledger/complete/wal/checkpointer.go

```diff
@@ -378,9 +401,13 @@ func StoreCheckpoint(writer io.Writer, tries ...*trie.MTrie) error {
 			if root == nil {
 				continue
 			}
```

Comment:
hi @fxamacker , could you explain why this `root` variable could possibly be `nil`?

Reply:
@zhangchiqing

> could you explain why this root variable could possibly be nil?

Empty trie is [always added to forest](https://github.com/onflow/flow-go/blob/c62e3ad9a198fd546c90cb2173720910fad620ff/ledger/complete/mtrie/forest.go#L47-L48) as starting point and empty trie's root is nil. It remains in the forest until evicted by trie queue exceeding capacity.

Maybe add this comment:

```suggestion
                        // Empty trie is always added to forest as starting point and 
                        // empty trie's root is nil. It remains in the forest until evicted
                        // by trie queue exceeding capacity.
			if root == nil {
				continue
			}
```

Reply:
Yeah, I think it makes sense [for root trie nodes](https://github.com/onflow/flow-go/pull/3139/files/77477eecee8ea809f31a8e8715bfe80088a209d4#diff-4d387a87aa55ecd3590e6cfeaa400f14d6aa9823fcbf4cded30e23c0117b767fR425-R427), but for subtrie roots, is it possible for them to be `nil`?

Reply:
You are right, the comment I suggested is for nil trie root (in a code block in different location).

Checking `root == nil` for subtries is useful for tests but it isn't likely to happen in non-test env.  Generated test data can be more sparse than mainnet's data.